kvbc/src/merkle_tree_db_adapter.cpp

```diff
@@ -448,48 +450,44 @@ Status DBAdapter::writeSTLinkTransaction(const Key &sTBlockKey, const Sliver &bl
     txn->put(key, value);
   }
 
-  try {
-    txn->commit();
-  } catch (const std::exception &e) {
-    const auto msg = std::string{"Failed to commit an addBlock() DB transaction, reason: "} + e.what();
-    LOG_ERROR(logger_, msg);
-    return Status::GeneralError(msg);
-  } catch (...) {
-    const auto msg = "Failed to commit an addBlock() DB transaction";
-    LOG_ERROR(logger_, msg);
-    return Status::GeneralError(msg);
-  }
-
-  return Status::OK();
+  // Commit the transaction.
+  txn->commit();
 }
 
-Status DBAdapter::addBlock(const Sliver &block, BlockId blockId) {
-  const auto lastReachableBlock = getLastReachableBlock();
+void DBAdapter::addRawBlock(const RawBlock &block, const BlockId &blockId) {
+  const auto lastReachableBlock = getLastReachableBlockId();
   if (blockId <= lastReachableBlock) {
     const auto msg = "Cannot add an existing block ID " + std::to_string(blockId);
     LOG_ERROR(logger_, msg);
-    return Status::IllegalOperation(msg);
+    throw std::invalid_argument{msg};
   } else if (lastReachableBlock + 1 == blockId) {
     // If adding the next block, append to the blockchain via the merkle tree and try to link with the ST temporary
     // chain.
-    const auto status = addLastReachableBlock(block::getData(block));
-    if (!status.isOK()) {
-      LOG_ERROR(logger_, "Failed to add a block to the end of the blockchain on state transfer, block ID " << blockId);
-      return status;
+    addBlock(block::getData(block));
+
+    try {
+      linkSTChainFrom(blockId + 1);
+    } catch (const std::exception &e) {
+      LOG_FATAL(logger_, "Aborting due to failure to link chains after block has been added, reason: "s + e.what());
+      std::exit(1);
```

Comment:
same as above about exit(1)

Reply:
This is a special case in that if we don't exit at the spot, we risk leaving the system in an invalid state, i.e. the block has been added, but the chains are not linked. If we just throw, we rely on the exception handler terminating the program and I prefer we be explicit here so that future code changes won't silently introduce a bug. When starting the system after the exit() call, the DBAdapter constructor will link the chains.